### PR TITLE
Lessons/Tutorials

### PR DESCRIPTION
**Pull request by ntoll**
I've created this PR so the work on tutorials / lessons gets some visibility in the wider project. All comments, constructive critique and ideas are most welcome.

For modifications let's correct the obvious mistakes in this PR, merge it and then update with later PRs.

### CODE REVIEW

File: docs/lessons/buttons.rst [hunk 1]
```diff
@@ -0,0 +1,188 @@
+Buttons
+-------
+
+So far we have created code that makes the device do something. This is called
+*output*. However, we also need the device to react to things. Such things are
+called *input*.
```
Comment (carlosperate) [inline docs/lessons/buttons.rst]: Obviously English is not my primary language, but `such things are` in plural and then `input` in singular doesn't feel quite right.

File: docs/lessons/buttons.rst [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+
+All this script does is sleep for ten thousand milliseconds (i.e. 10 seconds)
+and then scrolls the number of times you pressed button ``A``. That's it!
+
+While it's a pretty useless script, it introduces a couple of interesting new
+ideas:
+
+#. The ``sleep`` *function* will make the micro:bit sleep for a certain number
+   of milliseconds. If you want a pause in your program, this is how to do it.
+   A *function* is just like a *method*, but it isn't attached by a dot to an
```
Comment (carlosperate) [inline docs/lessons/buttons.rst]: Just to make sure, if we are defining functions based on the definition of methods, at this point it is assumed that the user is already familiar methods?

Comment (ntoll) [inline docs/lessons/buttons.rst]: Yes, that's introduced in an earlier "lesson" (see the index for the ordering).

File: docs/lessons/buttons.rst (hunk 1 above)
Comment (ntoll) [inline docs/lessons/buttons.rst]: Good catch. :-)

File: docs/lessons/buttons.rst [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+The third line is a bit like an onion. If the parenthesis are the
+onion skins then you'll notice that ``display.scroll`` contains ``str`` that
+itself contains ``button_a.get_presses``. Python attempts to work out the
+inner-most answer first before starting on the next layer out. This is called
+*nesting* - the coding equivalent of a Russian Matrioshka doll.
+
+.. image:: matrioshka.jpg
+
+Let's pretend you've pressed the button 10 times. Here's how Python works out
+what's happening on the third line:
```
Comment (carlosperate) [inline docs/lessons/buttons.rst]: `on the third line:` should that have a `::`?

File: docs/lessons/buttons.rst [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+than 10000 milliseconds (i.e. 10 seconds). If it is, *and this is where we can
+see scoping in action*, then it'll display ``Image.ASLEEP``. Notice how this is
+indented underneath the ``while`` statement *just like in our to-do list*.
+
+Obviously, if the running time is equal to or greater than 10000 milliseconds
+then the display will show ``Image.SURPRISED``. Why? Because the ``while``
+condition will be False (``running_time`` is no longer ``< 10000``). In that
+case the loop is finished and the program will continue after the ``while``
+loop's block of code. It'll look like your device is asleep for 10
+seonds before waking up with a surprised look on its face.
```
Comment (carlosperate) [inline docs/lessons/buttons.rst]: `seonds` to `seconds`

File: docs/lessons/buttons.rst (hunk 3 above)
Comment (ntoll) [inline docs/lessons/buttons.rst]: Not in this instance, because there is no code example that immediately follows.

File: docs/lessons/introduction.rst [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+
+These lessons introduce MicroPython and the BBC
+micro:bit in easy-to-follow steps. Feel free to adopt and adapt them for
+classroom based lessons, or perhaps just follow them on your own at home.
+
+You'll have most success if you explore, experiment and play. You can't break
+a BBC micro:bit by writing incorrect code. Just dive in!
+
+A word of warning: *you will fail many times*, and that is fine. **Failure is
+how good software developers learn**. Those of us who work professionally have
```
Comment (carlosperate) [inline docs/lessons/introduction.rst]: `Those of us who work professionally`
I can infer from the context that you mean work on software, but I would assume most people _should_ work professionally.

File: docs/lessons/buttons.rst (hunk 2 above)
Comment (carlosperate) [inline docs/lessons/buttons.rst]: Yes, I just noticed I was reading it in the wrong order. Started from the beginning now :)

File: docs/lessons/hello.rst [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+
+    display.scroll("Hello, World!")
+
+...tells MicroPython to use the display to scroll the string of characters
+"Hello, World!". The ``display`` part of that line is an *object* from the
+``microbit`` module that represents the device's physical display (we say
+"object" instead of "thingy", "whatsit" or "doodah"). We can tell the display
+to do things with a full-stop ``.`` followed by what looks like a command (in
+fact it's something we call a *method*). In this case we're using the
+``scroll`` method. Since ``scoll`` needs to know what characters to scroll
```
Comment (carlosperate) [inline docs/lessons/hello.rst]: ``scoll`` to ``scroll``

File: docs/lessons/buttons.rst (hunk 3 above)
Comment (carlosperate) [inline docs/lessons/buttons.rst]: Okay, I guess that showed my inexperience with rst, I'm more of a markdown guy myself.

File: docs/lessons/images.rst [hunk 7]
```diff
@@ -0,0 +1,234 @@
...
+DIY Images
+++++++++++
+
+Of course, you want to make your own image to display on the micro:bit, right?
+
+That's easy.
+
+Each LED pixel on the physical display can be set to one of ten values. If a
+pixel is set to ``0`` (zero) then it's off. It literally has zero brightness.
+However, if it is set to ``9`` then it is at its brightest level. The values
```
Comment (carlosperate) [inline docs/lessons/images.rst]: `if it is set to ``9`` then it is at its brightest level` to
`if it is set to ``9`` then it is at its highest brightest level`?

File: docs/lessons/introduction.rst (hunk 5 above)
Comment (ntoll) [inline docs/lessons/introduction.rst]: Good catch. :-)

File: docs/lessons/io.rst [hunk 8]
```diff
@@ -0,0 +1,91 @@
+Input/Output
+------------
+
+Along the bottom edge of the BBC micro:bit are strips of metal that make it
```
Comment (carlosperate) [inline docs/lessons/io.rst]: Should 'micro:bit are' be 'micro:bit there are'?

File: docs/lessons/music.rst [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+        tune = ["C4:4", "D", "E", "C", "C", "D", "E", "C", "E", "F", "G:8",
+                "E:4", "F", "G:8"]
+        music.play(tune)
+
+    Notice how the octave and duration values only change when they have to.
+
+Sound Effects
++++++++++++++
+
+It's possble to make tones that are not musical notes. For example, here's how
```
Comment (carlosperate) [inline docs/lessons/music.rst]: `possble` to `possible `

File: docs/lessons/music.rst [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+numeric values. The three arguments are the start value, end value and step
+size. Therefore, the first use of ``range`` is saying, in English, "create a
+range of values between 880 and 1760 in steps of 16". The second use of
+``range`` is saying, "create a range of values between 1760 and 880 in steps of
+-16".
+
+Because the siren should last forever it's wrapped in an infinite ``while``
+loop. Importantly, we have introduced a new sort of a loop inside the ``while``
+loop: the ``for`` loop. In English it's like saying, "for each item in some
+collection, do some thing". Specifically in the example above, it's saying,
```
Comment (carlosperate) [inline docs/lessons/music.rst]: should it be `do something`?

Comment (ntoll) [inline docs/lessons/music.rst]: Could be both but "something" is more common.

Keep these comments coming, they're *invaluable* and really helpful. As you've probably figured out, I'm writing all this during my commute in to/out of London, lunch time and in the evenings. It's hard to tell if it's making any sense...

:-)

File: docs/lessons/random.rst [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+Using the "onion" effect introduced when discussing *nesting* the
+``random.choice`` method takes this list as an argument and returns an item
+chosen at random. This item (the randomly chosen name) is the argument for
+``display.scroll``.
+
+Random Numbers
+++++++++++++++
+
+Random numbers are very useful. They're common in games. Why else do we have
+dice?
```
Comment (carlosperate) [inline docs/lessons/random.rst]: should it be `dices`?

Comment (tjguk) [inline docs/lessons/random.rst]: On 18/02/2016 12:24, Carlos wrote:
> In docs/lessons/random.rst
> <https://github.com/bbcmicrobit/micropython/pull/212#discussion_r53306783>:
> 
>> +
>> +    names = ["Amy", "Ben", "Damien", "Alia", "Kushal", "Sophie", "Zoltan" ]
>> +
>> +    display.scroll(random.choice(names))
>> +
>> +The list (``names``) contains seven names defined as strings of characters.
>> +Using the "onion" effect introduced when discussing *nesting* the
>> +``random.choice`` method takes this list as an argument and returns an item
>> +chosen at random. This item (the randomly chosen name) is the argument for
>> +``display.scroll``.
>> +
>> +Random Numbers
>> +++++++++++++++
>> +
>> +Random numbers are very useful. They're common in games. Why else do we have
>> +dice?
> 
> should it be |dices|?

Nope. The singular (seldom used) is die. The plural is dice.

File: docs/lessons/random.rst [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+``display.show`` expects a character then we use the ``str`` function to turn
+the numeric value into a character (we turn, for example, ``6`` into ``"6"``).
+
+If you know you'll always want a number between ``0`` and ``N`` then use the
+``random.randrange`` method. If you give it a single argument it'll return
+random integers up to, but not including, the value of the argument (this is
+different to the behaviour of ``random.randint``).
+
+Sometimes you need numbers with a decimal point in them. These are called
+*floating point* numbers and it's possible to such a number with the
```
Comment (carlosperate) [inline docs/lessons/random.rst]: `possible to such` -> `possible to generate such`

File: docs/lessons/random.rst [hunk 13]
```diff
@@ -0,0 +1,88 @@
...
+``random.random`` method. This only returns values between ``0.0`` and ``1.0``.
```
Comment (carlosperate) [inline docs/lessons/random.rst]: Maybe a good idea to specify if this version of random includes the values 0.0 and 1.0 themselves.

File: docs/lessons/random.rst [hunk 14]
```diff
@@ -0,0 +1,88 @@
...
+Seeds of Chaos
+++++++++++++++
+
+The random number generators used by computers are not truly random. They just
+give random like results given a starting *seed* value. The seed is often
+generated from random-ish values such as the current time and/or readings from
+sensors such as the thermometers built into chips.
+
+Sometimes you want to have repeatable random-ish behaviour: a source of
+randomness that is reproducable. It's like saying that you need the same five
```
Comment (carlosperate) [inline docs/lessons/random.rst]: `reproducable` -> `reproducible`

File: docs/lessons/random.rst (hunk 11 above)
Comment (carlosperate) [inline docs/lessons/random.rst]: Learn something new every day, thanks @tjguk!

File: docs/lessons/music.rst (hunk 10 above)
Comment (carlosperate) [inline docs/lessons/music.rst]: No problem. It looks pretty good so far, everything flows quite nicely and the concepts build on top each other quite well. It'll probably be more useful to have somebody with little programming knowledge to read through them, as they will really be able to notice if any of the concepts are hard to follow.

File: docs/lessons/direction.rst [hunk 15]
```diff
@@ -0,0 +1,27 @@
...
++++++++
+
+It can also tell you the direction of North like this::
+
+    from microbit import *
+
+    compass.calibrate()
+
+    while True:
+        needle = ((15 - compass.heading()) // 30) % 12
```
Comment (carlosperate) [inline docs/lessons/direction.rst]: The modulo and floor division operators have not really been mentioned here or anywhere else. While maybe that's outside of the scope for this documentation, it just feels a bit incomplete having these type of calculation here without any explanation.

File: docs/lessons/network.rst [hunk 16]
```diff
@@ -0,0 +1,298 @@
...
+
+    ...././.-../.-../---/.--/---/.-./.-../-..
+
+Can you work out what it says?
+
+Application
++++++++++++
+
+It's all very well having such a network stack, but you also need a way to
+interract with it - some form of application to send and recieve messages.
```
Comment (carlosperate) [inline docs/lessons/network.rst]: `interract` -> `interact`

File: docs/lessons/network.rst [hunk 17]
```diff
@@ -0,0 +1,298 @@
...
+With two layers in we have a physical connection and a description of how the
+signal will be sent. The next step involves describing how to handle a signal.
+For that we need a...
+
+Protocol
+++++++++
+
+If you ever meet the Queen there are expectations about how you ought to
+behave. For example, when she arrives you may bow or curtsey, if she offers her
+hand politely shake it, refer to her as "your mejesty" and thereafter as
```
Comment (carlosperate) [inline docs/lessons/network.rst]: `mejesty` -> `majesty`

File: docs/lessons/network.rst [hunk 18]
```diff
@@ -0,0 +1,298 @@
...
+"L" the signal is also switched on four times, but the second signal has a
+longer duration (``.-..``).
+
+Obviously, the timing of the signal is important: we need to tell a dot from a
+dash. That's another point of a protocol, to agree such things so everyone's
+implementation of the protocol will work with everyone elses. In this instance
+we'll just say that:
+
+* A signal with a duration less than 250 milliseconds is a dot.
+* A signal with a duration from 250 milliseconds to less than 500 milliseconds is a dash.
```
Comment (carlosperate) [inline docs/lessons/network.rst]: so what happens if a signal is longer than 500ms?

File: docs/lessons/direction.rst (hunk 15 above)
Comment (ntoll) [inline docs/lessons/direction.rst]: I'll add a clarification so those that are interested will be able to work out what's happening on this line, but I think it outside the scope of the documentation to explain it in detail. In fact, working out how it works *IS* quite good fun. ;-)

**Comment by ntoll** [comment]
Thank you @carlosperate - I've made modifications in light of your suggestions.

@dpgeorge care to merge this as a first draft?

**Comment by dpgeorge** [comment]
This is very comprehensive!

Are there any issues with licensing/copyright of the images (eg the queen)?

**Comment by ntoll** [comment]
All the images are from the Wikimedia commons and acknowledged as such as footnotes. There was but one exception: the gif of the breakfast robot (but I've emailed Simone to ask her permission yesterday). Since I've not heard back from her I've removed it and we can, if she's happy, re-add it.

**Comment by dpgeorge** [comment]
On thing to improve: for the image lesson, with the sinking boat, it's probably a good idea to introduce the image.shift_xxx methods so that you can make the images much more efficiently (in terms of writing code and also memory usage).

**Comment by ntoll** [comment]
@dpgeorge I did wonder about `shift_foo` for the boat but decided that the less efficient but more obvious-to-beginners method was the way to go. These are very much tutorials for the complete newbie teacher. ;-)

**Comment by dpgeorge** [comment]
I understand. I'm just worried that the person learning from this will
forevermore create animations this way :) I was thinking that we can add a
note at the end of the lesson pointing to the docs/api reference for the
Image class and mention that there's a much easier way to make the boat
sink (and it can be left as an exercise to the reader).

Also might be worth adding such pointers to the API reference in the other
lessons?

**Comment by ntoll** [comment]
@dpgeorge agreed. I think an extra tutorial called "Next Steps" should be added to the end to address these issues. It's 4am in the US and I'm wide awake so give me an hour and I'll rustle up a tutorial. :-)
